### PR TITLE
virt: use separate variable to hold TEE ram size

### PR DESCRIPTION
**Pull request by lorc**
VCORE_UNPG_RW_SZ is an linker symbol and it is affected by ASLR: when ASLR
code relocates image, it also affect values of all linker symbols.

As a workaround we shall deduct load offset from VCORE_UNPG_RW_SZ
to get actual TEE ram size. This change also adds clarity to the code.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

### CODE REVIEW

**Comment by lorc** [comment]
I'd prefer to merge this change before #4354.

In this case, I can drop patch "config.mk: force disable CORE_ASLR if VIRTUALIZATION is enabled" from #4354

**Comment by jforissier** [comment]
It looks like something is wrong in the linker script that defines `__flatmap_unpg_rw_size` and `__vcore_unpg_rw_size`. A size is a difference between two addresses so the symbol itself should not be relative (not subject to relocations). Would something like this help?
```diff
diff --git a/core/arch/arm/kernel/kern.ld.S b/core/arch/arm/kernel/kern.ld.S
index 50bde7a7..431016dc 100644
--- a/core/arch/arm/kernel/kern.ld.S
+++ b/core/arch/arm/kernel/kern.ld.S
@@ -295,7 +295,7 @@ SECTIONS
        }

        /* Start page aligned read-only memory */
-       __flatmap_unpg_rw_size = . - __flatmap_unpg_rw_start;
+       __flatmap_unpg_rw_size = ABSOLUTE(. - __flatmap_unpg_rw_start);

        __init_start = .;
        __flatmap_init_rx_start = .;
@@ -399,7 +399,7 @@ SECTIONS
        _end_of_ram = .;

 #ifndef CFG_WITH_PAGER
-       __flatmap_unpg_rw_size = _end_of_ram - __flatmap_unpg_rw_start;
+       __flatmap_unpg_rw_size = ABSOLUTE(_end_of_ram - __flatmap_unpg_rw_start);
        __get_tee_init_end = .;
 #endif

@@ -446,7 +446,7 @@ __vcore_unpg_ro_size = 0;

 /* Unpaged read-write memory */
 __vcore_unpg_rw_start = __flatmap_unpg_rw_start;
-__vcore_unpg_rw_size = __flatmap_unpg_rw_size;
+__vcore_unpg_rw_size = ABSOLUTE(__flatmap_unpg_rw_size);

 #ifdef CFG_VIRTUALIZATION
 /* Nexus read-write memory */
```

Not sure about the last one. I remember reading about the rules how the linker determines if a symbol should be relative ou absolute, it was quite complex and designed to work as expected in most cases... but perhaps here we need to be more persuasive.

Another thing, why is this not working for you when it does in QEMU etc.? What is the link with virtualization (`virt:`)? I could understand that different toolchains could behave differently though.

File: core/arch/arm/kernel/virtualization.c [hunk 1]
```diff
@@ -138,6 +140,12 @@ void virt_init_memory(struct tee_mmap_region *memory_map)
 {
 	struct tee_mmap_region *map;
 
+	/*
+	 * We can't use VCORE_UNPG_RW_SZ value directly, because it is altered
+	 * by ASLR.
+	 */
+	tee_ram_size = VCORE_UNPG_RW_SZ - boot_mmu_config.load_offset;
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/virtualization.c:147]: I guess that means that you haven't found a way either to avoid having `VCORE_*_SZ` altered by ASLR.

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by lorc** [comment]
> Would something like this help?

This was my first intention. But it does not work.

`static const size_t VCORE_UNPG_RW_SZ = ((size_t)__vcore_unpg_rw_size);` does not help either. On other work, it behaves well, if I make it global variable. But in this case `register_phys_mem_ul(MEM_AREA_TEE_RAM_RW, VCORE_UNPG_RW_PA, VCORE_UNPG_RW_SZ);` fails to compile, because it can't determine value of `VCORE_UNPG_RW_SZ` in compile time.

Anyways, my proposed patch is not the best solution. It fixes issue in `virtualization.c`, but I can see that `VCORE_UNPG_RW_SZ` is also used `mobj_init()`. I'm surprised, that `mobj_init()` does not suffer from a similar issue.

So, I'll try to find a better fix, that will benefit all parts of OP-TEE.

**Comment by lorc** [comment]
Closing this in favor of #4358, which is more generic and fixes issue with mobj as well.

**Pull request closed without merging**